### DISCUSSION TITLE
Unable to start an application with mcuboot using nucleo_h745zi_q board

### DISCUSSION THREAD

**Discussion by joelguittet** [zephyrproject-rtos/zephyr#62802] [Q&A] [answered]
Hello community !

I wonder how to create and start an application using mcuboot as a bootloader with the nucleo_h745zi_q board. I succeeded with no trouble with an other Nucleo board before but this is blocking for this one.

In order to show the issues I'm using the hello-world sample provided with mcuboot into https://github.com/mcu-tools/mcuboot/tree/main/samples/zephyr/hello-world

From this application, I have added the following overlays in the boards directory, to perform the configuration of the flash partition layout. just for information, the flash is separated with 2 banks (flash0 and flash1, flash1 that I would like to use later for the M4 core) and each bank has eight 128KB sectors:

nucleo_h745zi_q_m7_mcuboot.overlay:
```
/ {
    chosen {
        zephyr,code-partition = &boot_partition;
    };
};

&flash0 {
    partitions {
        compatible = "fixed-partitions";
        #address-cells = <1>;
        #size-cells = <1>;
        /* Bootloader slot: 128KB */
        boot_partition: partition@0 {
            label = "mcuboot";
            reg = <0x00000000 DT_SIZE_K(128)>;
            read-only;
        };
        /* Application slot: 384KB */
        slot0_partition: partition@20000 {
            label = "image-0";
            reg = <0x00020000 DT_SIZE_K(384)>;
        };
        /* Update slot: 384KB */
        slot1_partition: partition@80000 {
            label = "image-1";
            reg = <0x00080000 DT_SIZE_K(384)>;
        };
        /* Swap slot: 128KB */
        scratch_partition: partition@e0000 {
            label = "image-scratch";
            reg = <0x000E0000 DT_SIZE_K(128)>;
        };
    };
};
```

nucleo_h745zi_q_m7.overlay
```
/ {
    chosen {
        zephyr,code-partition = &slot0_partition;
    };
};


&flash0 {
    partitions {
        compatible = "fixed-partitions";
        #address-cells = <1>;
        #size-cells = <1>;
        /* Bootloader slot: 128KB */
        boot_partition: partition@0 {
            label = "mcuboot";
            reg = <0x00000000 DT_SIZE_K(128)>;
            read-only;
        };
        /* Application slot: 384KB */
        slot0_partition: partition@20000 {
            label = "image-0";
            reg = <0x00020000 DT_SIZE_K(384)>;
        };
        /* Update slot: 384KB */
        slot1_partition: partition@80000 {
            label = "image-1";
            reg = <0x00080000 DT_SIZE_K(384)>;
        };
        /* Swap slot: 128KB */
        scratch_partition: partition@e0000 {
            label = "image-scratch";
            reg = <0x000E0000 DT_SIZE_K(128)>;
        };
        /* Storage slot: 8KB */ //TODO
        /*storage_partition: partition@fe000 {
            label = "storage";
            reg = <0x000FE000 DT_SIZE_K(8)>;
        };*/
    };
};
```

Then building mcuboot with the following command:
```
west build -s $HOME/zephyrproject/bootloader/mcuboot/boot/zephyr -d build-mcuboot -b nucleo_h745zi_q_m7 -- -DDTC_OVERLAY_FILE=$HOME/zephyrproject/bootloader/mcuboot/samples/zephyr/hello-world/boards/nucleo_h745zi_q_m7_mcuboot.overlay -DCONFIG_BOOT_SWAP_USING_MOVE=y -DCONFIG_BOOT_MAX_IMG_SECTORS=8
```

Building and signing manually the hello-world application with the following:
```
west build -b nucleo_h745zi_q_m7 $HOME/zephyrproject/bootloader/mcuboot/samples/zephyr/hello-world
$HOME/zephyrproject/bootloader/mcuboot/scripts/imgtool.py sign --key $HOME/zephyrproject/bootloader/mcuboot/root-rsa-2048.pem --header-size 0x200 -S 0x60000 --align 8 --version "1.0" build/zephyr/zephyr.hex build/zephyr/zephyr-signed.hex
```

And flashing the target:
```
west flash -d build-mcuboot
west flash --hex-file build/zephyr/zephyr-signed.hex
```

mcuboot is starting but it seems the application is not, here are all the logs I get on the terminal:
```
*** Booting Zephyr OS build zephyr-v3.4.0-3926-g4ba97c22555e ***
I: Starting bootloader
I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: none
I: Image index: 0, Swap type: none
I: Bootloader chainload address offset: 0x20000
I: Jumping to the first image slot
```

Well and even worse, before that but I don't remember what I have changed, mcuboot was complaining about the validity of the application slot.... I haven't recorded the logs.

Is anybody getting the same issue ? Is there any settings related to the hardware, sector size, ... whatever, I should take care and I have not ?

Thanks!

Joel

PS : Zephyr is v3.4.0.

**Comment by nordicjm** [comment]
That sample in the mcuboot repository should not be used, it is for internal CI testing only. I would suggest getting `main` of zephyr and using sysbuild to build the bootloader and application, the flash entries can go in the dts in the board directory itself (no need for overlays and flash partitions for a real board/product should not be set in overlays) then you can build an application e.g. `samples/hello_world` by doing the following (whilst in the `samples/hello_world` directory):
```
mkdir build
cd build
cmake -GNinja -DBOARD=nucleo_h745zi_q -DAPP_DIR=.. -DSB_CONFIG_BOOTLOADER_MCUBOOT=y ../../share/sysbuild/
ninja
```
At this point, you will have both your bootloader and signed application hex files, and can then flash them to the board and see if it boots and works.

**Reply by joelguittet** [reply]
Hello @nordicjm 

Thanks for the information about this application. Understood.

Regarding to the commands you indicate, first this is the first time I see sysbuild stuff. Until now I have followed https://docs.zephyrproject.org/latest/services/device_mgmt/dfu.html#mcuboot and also https://docs.mcuboot.com/readme-zephyr that is linked in the zephyr documentation too. Perfectly working for an STM32L4 board. Your method do not follow at all the documentation so I'm a bit surprise to discover this. I would prefer something standard, using west probably.

Then, have you really tried this ? the cmake command above do not work at all for me. I started in `samples/hello_world` directory and first I notice path to share sysbuild dir is wrong, there is a missing `../`. Even fixing this, this is not building. There are probably missing variables.

Joel

**Reply by mariopaja** [reply, marked as answer]
Hello @joelguittet 

Sure you can use west. In this case, I have copied the `hello_world` sample at `~/zephyrproject/` and added the following in the `prj.conf`.

```
# Print a banner on the UART on startup.
CONFIG_BOOT_BANNER=y

# Enable console and printk()
CONFIG_PRINTK=y
CONFIG_STDOUT_CONSOLE=y

# Enable Zephyr application to be booted by MCUboot
CONFIG_BOOTLOADER_MCUBOOT=y
```

The only commands needed now:

`west build -p -b nucleo_h745zi_q_m7 -s ~/zephyrproject/bootloader/mcuboot/boot/zephyr -- -DCONFIG_BOOT_SWAP_USING_MOVE=y -DCONFIG_BOOT_MAX_IMG_SECTORS=8`

`west flash --hex-file ~/zephyrproject/bootloader/mcuboot/boot/zephyr/build/zephyr/zephyr.hex`

`west build -p -b nucleo_h745zi_q_m7 -s ~/zephyrproject/hello_world`

`west sign -t imgtool -d ~/zephyrproject/hello_world/build -- --version 1.0.0 --pad --key bootloader/mcuboot/root-rsa-2048.pem`

`west flash --hex-file ~/zephyrproject/hello_world/build/zephyr/zephyr.signed.hex`

BTW, Thank you for the overlay. I was trying the past days and I was struggling a bit :)


```
*** Booting Zephyr OS build zephyr-v3.4.0-3831-g1731010869c4 ***
I: Starting bootloader
I: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: primary slot
I: Image index: 0, Swap type: none
I: Bootloader chainload address offset: 0x20000
I: Jumping to the first image slot
*** Booting Zephyr OS build zephyr-v3.4.0-3831-g1731010869c4 ***
Hello World! nucleo_h745zi_q_m7
```

**Reply by joelguittet** [reply]
Hello @mariopaja 

> I have copied the hello_world sample at ~/zephyrproject/ and changed the prj.conf

What have you changed in prj.conf please ?

Thanks

**Reply by mariopaja** [reply]
@joelguittet 

I added these in prj.conf because I used project from zephyr/samples/hello_world

```
# Print a banner on the UART on startup.
CONFIG_BOOT_BANNER=y

# Enable console and printk()
CONFIG_PRINTK=y
CONFIG_STDOUT_CONSOLE=y

# Enable Zephyr application to be booted by MCUboot
CONFIG_BOOTLOADER_MCUBOOT=y
```

**Reply by joelguittet** [reply]
Thanks for sharing this !!!
Working for me as well, I was not using the `west sign` command, but instead calling the script myself, and it seems the issue was there !
Indicated your message "mark as answer" :)
